ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/SubordinationManagerXAResourceOrphanFilter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2010, Red Hat, Inc. and/or its affiliates,
+ * and individual contributors as indicated by the @author tags.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2016,
+ * @author JBoss, by Red Hat.
+ */
+package com.arjuna.ats.internal.jta.recovery.arjunacore;
+
+import com.arjuna.ats.arjuna.coordinator.TxControl;
+import com.arjuna.ats.arjuna.recovery.RecoveryManager;
+import com.arjuna.ats.arjuna.recovery.RecoveryModule;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinateTransaction;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinationManager;
+import com.arjuna.ats.jta.recovery.XAResourceOrphanFilter;
+import com.arjuna.ats.jta.xa.XATxConverter;
+import com.arjuna.ats.jta.xa.XidImple;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.Xid;
+import java.util.Vector;
+
+/**
+ * An XAResourceOrphanFilter which vetos rollback for xids which have an in-flight subordinate transaction.
+ * <p>
+ * The SubordinateAtomicActionRecoveryModule must be loaded in in a position prior to the XARecoveryModule for this to work
+ */
+public class SubordinationManagerXAResourceOrphanFilter implements XAResourceOrphanFilter {
+    private boolean validated;
+
+    @Override
+    public Vote checkXid(Xid xid) {
+        if (xid.getFormatId() != XATxConverter.FORMAT_ID) {
+            // we only care about Xids created by the JTA
+            return Vote.ABSTAIN;
+        }
+
+        if (!TxControl.getXANodeName().equals(XATxConverter.getSubordinateNodeName(new XidImple(xid).getXID()))) {
+            // It either doesn't have a subordinate node name or isn't for this server
+            return Vote.ABSTAIN;
```

Comment:
could happen that the `XATxConverter.getSubordinateNodeName` to `getXID` will not match but the `NodeNameXAResourceOrphanFilter` will match the `XATxConverter.getNodeName` for the same XID.

This would mean - this filter votes for {{ABSTAIN}} but the orphan filter then votes for rollback.

I can't find some situation to reach this from top of my head. Just I would like to mention such possibility.

Reply:
I see what you are saying: "what if the parent recovery manager also has access to the same resource manager and scans a shared XAResource - won't it try to roll it back"? I will add an extra check into NodeNameXAResourceOrphanFilter to make sure that the original NodeNameXAResourceOrphanFilter does not do that

Reply:
That was not necessary eventually as only the node that is ultimately responsible for NodeNameXAResourceOrphanFilter will match the XATxConverter.getNodeName and that has the further orphan check that looks in the ActionStatusService